### ISSUE TITLE
onnx::NonZero (torch.nonzero) operation not supported

### ISSUE THREAD

**Issue by muaz-urwa** [openvinotoolkit/openvino#435]

**Comment by muaz-urwa** [comment]
I am testing with a simple pytorch network: 

class test_model(nn.Module):

    def __init__(self, in_channels, representation_size):
        super(test_model, self).__init__()

        self.fc6 = nn.Linear(in_channels, representation_size)

    def forward(self, x):
        x = x.flatten(start_dim=1)
        x = F.relu(self.fc6(x))
        x = torch.nonzero(x)
        return x

Converting it to onnx. Onnx log is below:
graph(%0 : Float(1, 50),
      %fc6.weight : Float(5, 50),
      %fc6.bias : Float(5)):
  %3 : Float(1, 50) = onnx::Flatten[axis=1](%0) # <ipython-input-4-ad4b4b8cd54f>:9:0
  %4 : Float(1, 5) = onnx::Gemm[alpha=1, beta=1, transB=1](%3, %fc6.weight, %fc6.bias) # /opt/conda/lib/python3.7/site-packages/torch/nn/functional.py:1370:0
  %5 : Float(1, 5) = onnx::Relu(%4) # /opt/conda/lib/python3.7/site-packages/torch/nn/functional.py:914:0
  %6 : Tensor = onnx::NonZero(%5)
  %7 : Long(2, 2) = onnx::Transpose[perm=[1, 0]](%6) # <ipython-input-4-ad4b4b8cd54f>:12:0
  return (%7)

When I try to create IR files from it. I get following error depending on openvino version.

**openvino_2020.1.023:**
[ WARNING ]  Please set `version` attribute for node 6 with type=<UNKNOWN>
[ ERROR ]  List of operations that cannot be converted to Inference Engine IR:
[ ERROR ]      NonZero (1)
[ ERROR ]          6
[ ERROR ]  Part of the nodes was not converted to IR. Stopped.

**openvino_2019.3.376**
Model Optimizer version: 	2019.3.0-408-gac8584cb7
[ ERROR ]  Cannot infer shapes or values for node "6".
[ ERROR ]  There is no registered "infer" function for node "6" with op = "NonZero". Please implement this function in the extensions. 
 For more information please refer to Model Optimizer FAQ (https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_Model_Optimizer_FAQ.html), question #37. 
[ ERROR ]  
[ ERROR ]  It can happen due to bug in custom shape infer function <UNKNOWN>.
[ ERROR ]  Or because the node inputs have incorrect values/shapes.
[ ERROR ]  Or because input shapes are incorrect (embedded to the model or passed via --input_shape).
[ ERROR ]  Run Model Optimizer with --log_level=DEBUG for more information.
[ ERROR ]  There is no registered "infer" function for node "6" with op = "NonZero". Please implement this function

I using this command for conversion:
python3 mo_onnx.py --input_model \
/app/temp/test_model.onnx \
--model_name test_model_ir \
--data_type=FP32 \
--output_dir /app/temp/ \
--input_shape [1,50] \

**Model converts without torch.nonzero.**

**Comment by jgespino** [comment]
Hi @muaz-urwa 

Thanks for reaching out, the NonZero layer is not supported by the OpenVINO Toolkit. You can find a list of supported layers in the [Supported Framework Layers](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_Supported_Frameworks_Layers.html#onnx_supported_operators) document.

Regards,
Jesus

**Comment by muaz-urwa** [comment]
Hello Jesus. 

Yes, it is not mentioned it the supported layers list. But turns out that it is supported in 2020 release. 
Here is a link to the op file: [link](https://github.com/opencv/dldt/blob/b2140c083a068a63591e8c2e9b5f6b240790519d/model-optimizer/extensions/ops/non_zero.py)

But it only works if you add this flag to model optimizer call (mo.py):
--input "0:2"

as done [here](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_onnx_specific_Convert_Mask_RCNN.html)

**Comment by jgespino** [comment]
Hi @muaz-urwa 

Could you share your onnx model for me to take a quick look?

Regards,
Jesus

**Comment by muaz-urwa** [comment]
By onnx model you mean the onnx file ?

I was able to convert it into openvino IR successfully after using the --input flag as described above.

**Comment by jgespino** [comment]
Hi @muaz-urwa 

Yes, I meant your onnx file (test_model.onnx). Glad you were able to convert your model to IR format.
Let me know if you need further assistance.

Regards,
Jesus

**Comment by muaz-urwa** [comment]
Thank You.

I am closing the issue.

**Comment by muaz-urwa** [comment]
Hello Jesus,

Here is the [link](https://drive.google.com/file/d/1TGO6B_iPudltKHA8XNESXXydSGDye05K/view?usp=sharing) to onnx file.

I am still facing issues converting this model.

############################################
[mo.py] Starting Conversion to IR
Model Optimizer arguments:
Common parameters:
	- Path to the Input Model: 	/app/temp/test_model.onnx
	- Path for generated IR: 	/app/temp/
	- IR output name: 	test_model_ir
	- Log level: 	ERROR
	- Batch: 	Not specified, inherited from the model
	- Input layers: 	0
	- Output layers: 	Not specified, inherited from the model
	- Input shapes: 	Not specified, inherited from the model
	- Mean values: 	Not specified
	- Scale values: 	Not specified
	- Scale factor: 	Not specified
	- Precision of IR: 	FP32
	- Enable fusing: 	True
	- Enable grouped convolutions fusing: 	True
	- Move mean values to preprocess section: 	False
	- Reverse input channels: 	False
ONNX specific parameters:
Model Optimizer version: 	2020.1.0-61-gd349c3ba4a
[ WARNING ]  Please set `version` attribute for node 6 with type=<UNKNOWN>
[ ERROR ]  List of operations that cannot be converted to Inference Engine IR:
[ ERROR ]      NonZero (1)
[ ERROR ]          6
[ ERROR ]  Part of the nodes was not converted to IR. Stopped. 
 For more information please refer to Model Optimizer FAQ (https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_Model_Optimizer_FAQ.html), question #24.

**Comment by jgespino** [comment]
Hi @muaz-urwa 

I took a look at the model and the NonZero layer is currently not supported. The layer operation ([non_zero.py](https://github.com/opencv/dldt/blob/b2140c083a068a63591e8c2e9b5f6b240790519d/model-optimizer/extensions/ops/non_zero.py)) is included as you mentioned. However, the layer attribute extraction is still needed to make this custom layer operational.

Take a look at the [Custom Layers in Model Optimizer](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_customize_model_optimizer_Customize_Model_Optimizer.html) documentation for additional information on how to implement. Additionally, there is a YouTube series on [Custom Layers for OpenVINO](https://www.youtube.com/watch?v=UrfWgFKjaPc) that may be useful.

Regards,
Jesus

**Comment by muaz-urwa** [comment]
Dear Jesus,

Thank you for the guidance. 
I have two more questions:
1) Is there a way to find out if the layer is implemented in Inference Engine, because additions in Model optimizer are relatively simpler and can be done easily.
2) Latest release notes state that mask rcnn is now available. Here is a [tutorial](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_onnx_specific_Convert_Mask_RCNN.html) on how to convert it from onnx to openvino. Now Mask rcnn implementation has nonzero operation on multiple places. If you download the onnx file from link above it has multiple nonzero nodes. But following the tutorial model optimizer is able to convert it into IR. How is this happening if nonzero is not supported ?

**Comment by jgespino** [comment]
Hi @muaz-urwa 

That is a very good question! The mask_rcnn in the link you provided does have NonZero operations and converts successfully. I looked at the transformation configuration used in the guide and it doesn't look like NonZero operations are transformed. I will need to reach out to the development team and find out more. 

Regards,
Jesus

**Comment by lazarevevgeny** [comment]
The transformation configuration file triggers replacement of huge sub-graphs of the MaskRCNN to be replaced with just several operations ExperimentalDetectron*. These sub-graphs being replaced contain all NonZero operations in the model. Therefore there are no NonZero operations in the generated IR.

Currently NonZero operation is not supported in the DLDT.

**Comment by muaz-urwa** [comment]
Dear @lazarevevgeny,

Thank you for the explanation. I looked into these transformations and subgraph replacements for Experimental Detectrons. I was able to convert my model into IR by writing a similar conversion class for my model. But I have following reservations:

1. Every time I tweak my pytorch model a little bit, I have to go and manually read the entire onnx graph and note the nodes.
2. These input, output nodes and parameters of modules like ExperimentalDetectronRoiFeatureExtractor are harcoded into the conversion.

If the individual component nodes/operators were supported it would be able to automate the conversion across model architecture changes (which are frequent in my experiments). And I believe there are only two additional simple nodes/operators that need to be supported for faster rcnn. NonZero and RoiAlign.

What is your roadmap to include support for these operators ?
If you don't have any immediate plans, how would you entertain a pull requests for these features.

Regards,
Urwa

**Comment by muaz-urwa** [comment]
Waiting for response o the above.
Thank you.

**Comment by lazarevevgeny** [comment]
@muaz-urwa, unfortunately I cannot provide exact timeline. The is a significant technical obstacle to support NonZero operation - it has a dynamic output shape meaning that the output tensor shape depends on input values and not just input shape and operation attributes. The plugin teams are working on enabling this functionality but as I said I cannot provide any timeline.

**Comment by andrei-kochin** [comment]
@muaz-urwa it seems that according to point raised here all the issues supposed to be resolved on latest master. Unfortunately I can't access onnx file you've shared anymore as link is no more valid. Could you please verify that there is no issues now?

**Comment by jgespino** [comment]
Closing, please re-open if additional assistance is needed.
